File: test/e2e/Apps/BasicDotNetIsolated/OrchestrationQuery.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Net;
using Grpc.Core;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.DurableTask.Client;

namespace Microsoft.Azure.Durable.Tests.E2E;
public static class OrchestrationQueryFunctions
{
    [Function(nameof(GetAllInstances))]
    public static async Task<HttpResponseData> GetAllInstances(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableTaskClient client)
    {
        try 
        {
            var instances = client.GetAllInstancesAsync();

            var response = req.CreateResponse(HttpStatusCode.OK);
            await response.WriteAsJsonAsync(instances);
            return response;
        }
        catch (RpcException ex) 
        {
            var response = req.CreateResponse(HttpStatusCode.BadRequest);
            response.Headers.Add("Content-Type", "text/plain");
            await response.WriteStringAsync(ex.Message);
            return response;
        }
    }

    [Function(nameof(GetRunningInstances))]
    public static async Task<HttpResponseData> GetRunningInstances(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableTaskClient client)
    {
        try 
        {
            OrchestrationQuery filter = new OrchestrationQuery(Statuses: new List<OrchestrationRuntimeStatus> { 
                OrchestrationRuntimeStatus.Running, 
                OrchestrationRuntimeStatus.Pending, 
                OrchestrationRuntimeStatus.Suspended 
            });
            var instances = client.GetAllInstancesAsync(filter);

            var response = req.CreateResponse(HttpStatusCode.OK);
            await response.WriteAsJsonAsync(instances);
            return response;
        }
        catch (RpcException ex) 
        {
            var response = req.CreateResponse(HttpStatusCode.BadRequest);
            response.Headers.Add("Content-Type", "text/plain");
            await response.WriteStringAsync(ex.Message);
            return response;
        }
    }
}

File: test/e2e/Apps/BasicJava/src/main/java/com/function/OrchestrationQuery.java
package com.function;

import com.microsoft.azure.functions.annotation.*;
import com.function.JsonHelpers.DurableMetadataGsonProvider;
import com.microsoft.azure.functions.*;

import java.util.*;

import com.microsoft.durabletask.*;
import com.microsoft.durabletask.azurefunctions.DurableClientContext;
import com.microsoft.durabletask.azurefunctions.DurableClientInput;


public class OrchestrationQuery {
    /**
     * This HTTP-triggered function returns all orchestration instances.
     */
    @FunctionName("GetAllInstances")
    public HttpResponseMessage getAllInstances(
            @HttpTrigger(name = "req", methods = {HttpMethod.GET, HttpMethod.POST}, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<String>> request,
            @DurableClientInput(name = "durableContext") DurableClientContext durableContext,
            final ExecutionContext context) {
        DurableTaskClient client = durableContext.getClient();
        try {
            // Java SDK: getAllInstancesAsync returns a CompletableFuture<List<OrchestrationInstanceStatus>>
            OrchestrationStatusQueryResult instances = client.queryInstances(new OrchestrationStatusQuery());
            String instanceString = DurableMetadataGsonProvider.createGson().toJson(instances.getOrchestrationState());
            return request.createResponseBuilder(HttpStatus.OK)
                    .header("Content-Type", "application/json")
                    .body(instanceString)
                    .build();
        } catch (Exception ex) {
            context.getLogger().severe("Error querying all instances: " + ex.getMessage());
            return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                    .header("Content-Type", "text/plain")
                    .body(ex.getMessage())
                    .build();
        }
    }

    /**
     * This HTTP-triggered function returns running, pending, and suspended orchestration instances.
     */
    @FunctionName("GetRunningInstances")
    public HttpResponseMessage getRunningInstances(
            @HttpTrigger(name = "req", methods = {HttpMethod.GET, HttpMethod.POST}, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<String>> request,
            @DurableClientInput(name = "durableContext") DurableClientContext durableContext,
            final ExecutionContext context) {
        DurableTaskClient client = durableContext.getClient();
        try {
            List<OrchestrationRuntimeStatus> statuses = Arrays.asList(
                    OrchestrationRuntimeStatus.RUNNING,
                    OrchestrationRuntimeStatus.PENDING,
                    OrchestrationRuntimeStatus.SUSPENDED
            );
            OrchestrationStatusQuery query = new OrchestrationStatusQuery();
            query.setRuntimeStatusList(statuses);
            OrchestrationStatusQueryResult instances = client.queryInstances(query);
            String instanceString = DurableMetadataGsonProvider.createGson().toJson(instances.getOrchestrationState());
            return request.createResponseBuilder(HttpStatus.OK)
                    .header("Content-Type", "application/json")
                    .body(instanceString)
                    .build();
        } catch (Exception ex) {
            context.getLogger().severe("Error querying running instances: " + ex.getMessage());
            return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                    .header("Content-Type", "text/plain")
                    .body(ex.getMessage())
                    .build();
        }
    }
}